From e32622a9af583b1cf73555fb985b49e0f73223f7 Mon Sep 17 00:00:00 2001
From: Dmitry Muhomor <muhomor.dmitry@gmail.com>
Date: Mon, 28 Aug 2023 15:51:24 +0300
Subject: [PATCH] aapt2: add stringified resource configuration to
 Configuration.proto

This is needed for adevtool, to reuse large ConfigDescription.to_string() function which is used
for describing resource qualifiers, e.g. "values-mcc123-sw600dp-night-en".
---
 tools/aapt2/Configuration.proto             | 4 +++-
 tools/aapt2/cmd/Convert.h                   | 3 +++
 tools/aapt2/format/proto/ProtoSerialize.cpp | 1 +
 3 files changed, 7 insertions(+), 1 deletion(-)

diff --git a/tools/aapt2/Configuration.proto b/tools/aapt2/Configuration.proto
index f27dd1d831726..f417612ce3c77 100644
--- a/tools/aapt2/Configuration.proto
+++ b/tools/aapt2/Configuration.proto
@@ -217,5 +217,7 @@ message Configuration {
 
   string product = 25;
 
-  reserved 28 to max;  // Next IDs
+  string stringified = 28;
+
+  reserved 29 to max;  // Next IDs
 }
diff --git a/tools/aapt2/cmd/Convert.h b/tools/aapt2/cmd/Convert.h
index 60669b08a15f0..e88795b41477d 100644
--- a/tools/aapt2/cmd/Convert.h
+++ b/tools/aapt2/cmd/Convert.h
@@ -76,6 +76,8 @@ class ConvertCommand : public Command {
         "Has no effect on APKs where resource names are kept.",
         &table_flattener_options_.deduplicate_entry_values);
     AddOptionalSwitch("-v", "Enables verbose logging", &verbose_);
+    AddOptionalSwitch("--for-adevtool", "This flag exists to prevent using adevtool with "
+            "upstream aapt2 which lacks the Configuration.proto customization.", &for_adevtool_);
   }
 
   int Action(const std::vector<std::string>& args) override;
@@ -93,6 +95,7 @@ class ConvertCommand : public Command {
   bool force_sparse_encoding_ = false;
   bool enable_compact_entries_ = false;
   std::optional<std::string> resources_config_path_;
+  bool for_adevtool_ = false;
 };
 
 int Convert(IAaptContext* context, LoadedApk* input, IArchiveWriter* output_writer,
diff --git a/tools/aapt2/format/proto/ProtoSerialize.cpp b/tools/aapt2/format/proto/ProtoSerialize.cpp
index 6e0cf6f80630c..f981cf7a53ee8 100644
--- a/tools/aapt2/format/proto/ProtoSerialize.cpp
+++ b/tools/aapt2/format/proto/ProtoSerialize.cpp
@@ -63,6 +63,7 @@ static pb::Visibility::Level SerializeVisibilityToPb(Visibility::Level state) {
 }
 
 void SerializeConfig(const ConfigDescription& config, pb::Configuration* out_pb_config) {
+  out_pb_config->set_stringified(config.to_string());
   out_pb_config->set_mcc(config.mcc);
   out_pb_config->set_mnc(config.mnc);
   out_pb_config->set_locale(config.GetBcp47LanguageTag());

